### PR TITLE
feat: add CombinedServices for parsing netv1.Ingress

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Adds a default off `CombinedServices` feature gate that will prevent IngressV1 translator from creating multiple services when multiple ingresses use the same service as their backend.

**Which issue this PR fixes**:

Solves #3777.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/parser/translate_ingress_test.go [hunk 1]
```diff
@@ -1102,6 +1102,195 @@ func TestFromIngressV1(t *testing.T) {
 			assert.Equal(t, *parsedInfo.ServiceNameToServices["foo-namespace.foo-2.foo-svc.80"].Host, "foo-svc.foo-namespace.80.svc")
 		})
 	})
+
+	t.Run("CombinedRoutes=on && CombinedServices=on", func(t *testing.T) {
+		setupParser := func(t *testing.T, store store.Storer) *Parser {
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_ingress_test.go:1107]: can we use a table driven test structure here instead? This could make the file shorter. (maybe we could do this in a separate refactoring PR)

File: CHANGELOG.md [hunk 2]
```diff
@@ -88,6 +88,18 @@ Adding a new version? You'll need three changes:
...
+- Added `CombinedServices` feature gate that prevents the controller from
+  creating a separate Kong `Service` for each `netv1.Ingress` that uses
+  the same Kubernetes `Service` as its backend when `CombinedRoutes` feature
+  is turned on. Instead, the controller will create a single Kong `Service`
+  in such circumstances.
+  The feature is disabled by default to allow users to adapt to the changed
+  Kong Services naming scheme (`<namespace>.<service-name>.<port>` instead
+  of `<namespace>.<ingress-name>.<service-name>.<port>`) - it could break
+  existing monitoring rules that rely on the old naming scheme.
+  It will become the default behavior in the next minor release with the possibility
```
Comment (pmalek) [inline CHANGELOG.md:100]:
```suggestion
  It will become the default behavior in the next major release with the possibility
```

?

File: internal/dataplane/parser/translate_ingress.go [hunk 3]
```diff
@@ -206,23 +206,18 @@ func (p *Parser) ingressRulesFromIngressV1() ingressRules {
...
 		result.SecretNameToSNIs.addFromIngressV1TLS(ingressSpec.TLS, ingress)
-
-		if wasServicesCacheUpdated := p.ingressV1ToKongService(ingress, icp, servicesCache); wasServicesCacheUpdated {
-			p.reportKubernetesObjectUpdate(ingress)
-		}
 	}
 
-	for _, service := range servicesCache {
+	// Translate Ingress objects into Kong Services.
+	for _, service := range p.ingressesV1ToKongServices(ingressList, icp) {
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_ingress.go:220]: Can we add a comment to tell that the parent could be non-unique, but they should be in the same namespace because ingress can only refer to services in the same namespace?

File: internal/manager/featuregates/feature_gates.go [hunk 4]
```diff
@@ -33,6 +33,10 @@ const (
 	// See: https://docs.konghq.com/gateway/latest/key-concepts/routes/expressions/ about expression based routes in Kong 3.0+.
 	ExpressionRoutesFeature = "ExpressionRoutes"
 
+	// CombinedServicesFeature is the name of the feature-gate that makes KIC create a single Kong Service when
+	// a Kubernetes Service is referenced by multiple netv1.Ingresses. It's effective only when CombinedRoutesFeature is enabled.
+	CombinedServicesFeature = "CombinedServices"
```
Comment (randmonkey) [inline internal/manager/featuregates/feature_gates.go:38]: Seems that the PR did not add the code to apply the feature gate. Are you going to leave this for another PR?

File: CHANGELOG.md (hunk 2 above)
Comment (czeslavo) [reply CHANGELOG.md:100]: That's what we agreed upon when discussing the strategy for rolling such changes out at our team meeting: https://docs.google.com/document/d/1Ct-ZOceuXFUISPWrzQG8YMRCGPSVDe7RTdVQZ6uzq_U/edit#heading=h.1iupcjhqk605
```
Minor X -> off by default
Minor X+1 -> on by default, possibility to opt-out
Minor X+2 -> on by default, feature gate removal
```

File: internal/dataplane/parser/objects.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package parser
+
+import "sigs.k8s.io/controller-runtime/pkg/client"
+
+// ObjectsCollector collects objects for later use.
+// Its methods are safe to call with a nil receiver.
+type ObjectsCollector struct {
```
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/dataplane/parser/objects.go:7]: It's used only in a single-threaded code path in the Parser, so there's no risk of concurrent access.

File: internal/dataplane/parser/translate_ingress.go (hunk 3 above)
Comment (czeslavo) [reply internal/dataplane/parser/translate_ingress.go:220]: I'm not sure if I understand your comment correctly: should I put a comment explaining the `service.Parent` uniqueness and its namespace association just above the `result.ServiceNameToParent[*service.Name] = service.Parent` line? 🤔 IMO it's not the best place to put that, as the parent assignment originally happens somewhere else, here it's just "repacked" to the result. How about adding a comment above `Parent` field in the `kongstate.Service` struct? a6f9a53

File: internal/manager/featuregates/feature_gates.go (hunk 4 above)
Comment (czeslavo) [reply internal/manager/featuregates/feature_gates.go:38]: That's a fair point, thanks for catching this 😅 a6f9a53

File: internal/dataplane/parser/translate_ingress_test.go (hunk 1 above)
Comment (czeslavo) [reply internal/dataplane/parser/translate_ingress_test.go:1107]: Yes, I agree that would be useful 👍 Let's tackle it separately: https://github.com/Kong/kubernetes-ingress-controller/issues/3978

**Comment by pmalek** [comment]
1 thing I'd consider is to maybe add integration tests matrix entry until this becomes enabled by default WDYT?

We would need an issue attached to subsequent minor release which would remove this entry because it would become redundant (given that it would be enabled by default).

**Comment by czeslavo** [comment]
> 1 thing I'd consider is to maybe add integration tests matrix entry until this becomes enabled by default WDYT?
> 
> We would need an issue attached to subsequent minor release which would remove this entry because it would become redundant (given that it would be enabled by default).

Added a matrix entry and created an issue to track its removal: https://github.com/Kong/kubernetes-ingress-controller/issues/3979.

**Review by pmalek — APPROVED** [review]

**Review by randmonkey — APPROVED** [review]
